driver-core/src/main/java/com/datastax/driver/core/SessionManager.java

```diff
@@ -220,68 +202,83 @@ ListeningExecutorService blockingExecutor() {
         if (distance == HostDistance.IGNORED)
             return Futures.immediateFuture(true);
 
-        // Creating a pool is somewhat long since it has to create the connection, so do it asynchronously.
-        return executor.submit(new Callable<Boolean>() {
-            public Boolean call() {
-                while (true) {
-                    try {
-                        if (isClosing)
-                            return true;
+        if (isClosing)
+            return Futures.immediateFuture(false);
 
-                        HostConnectionPool newPool = new HostConnectionPool(host, distance, SessionManager.this);
-                        HostConnectionPool previous = pools.put(host, newPool);
-                        if (previous == null) {
-                            logger.debug("Added connection pool for {}", host);
-                        } else {
-                            logger.debug("Renewed connection pool for {}", host);
-                            previous.closeAsync();
-                        }
+        final HostConnectionPool newPool = new HostConnectionPool(host, distance, SessionManager.this);
+        newPool.initAsync();
 
-                        // If we raced with a session shutdown, ensure that the pool will be closed.
-                        if (isClosing)
-                            newPool.closeAsync();
+        final SettableFuture<Boolean> future = SettableFuture.create();
 
-                        return true;
-                    } catch (Exception e) {
-                        logger.error("Error creating pool to " + host, e);
-                        return false;
-                    }
+        Futures.addCallback(newPool.initFuture, new FutureCallback<Void>() {
+            @Override
+            public void onSuccess(Void result) {
+                // If we raced with a session shutdown, ensure that the pool will be closed.
+                HostConnectionPool previous = pools.put(host, newPool);
+                if (previous == null) {
+                    logger.debug("Added connection pool for {}", host);
+                } else {
+                    logger.debug("Renewed connection pool for {}", host);
+                    previous.closeAsync();
                 }
+
+                // If we raced with a session shutdown, ensure that the pool will be closed.
+                if (isClosing) {
+                    newPool.closeAsync();
+                    pools.remove(host);
+                    future.set(false);
+                } else {
+                    future.set(true);
+                }
+            }
+            @Override
+            public void onFailure(Throwable t) {
+                logger.error("Error creating pool to " + host, t);
+                future.set(false);
             }
         });
```

Comment:
We are not using the blocking executor here, and the parameter `executor` could be removed.